mm2src/coins/nft/storage/wasm/wasm_storage.rs

```diff
@@ -28,144 +31,517 @@ impl IndexedDbNftStorage {
-        _filters: Option<NftTxHistoryFilters>,
+        chains: Vec<Chain>,
+        max: bool,
+        limit: usize,
+        page_number: Option<NonZeroUsize>,
+        filters: Option<NftTxHistoryFilters>,
     ) -> MmResult<NftsTransferHistoryList, Self::Error> {
-        todo!()
+        let locked_db = self.lock_db().await?;
+        let db_transaction = locked_db.get_inner().transaction().await?;
+        let table = db_transaction.table::<NftTxHistoryTable>().await?;
+        let mut txs = Vec::new();
+        for chain in chains {
+            let tx_tables = table
+                .get_items("chain", chain.to_string())
+                .await?
+                .into_iter()
+                .map(|(_item_id, tx)| tx);
+            let filtered = Self::take_txs_according_to_filters(tx_tables, filters)?;
+            txs.extend(filtered);
+        }
+        Self::take_txs_according_to_paging_opts(txs, max, limit, page_number)
     }
 
-    async fn add_txs_to_history<I>(&self, _chain: &Chain, _txs: I) -> MmResult<(), Self::Error>
+    async fn add_txs_to_history<I>(&self, _chain: &Chain, txs: I) -> MmResult<(), Self::Error>
     where
         I: IntoIterator<Item = NftTransferHistory> + Send + 'static,
         I::IntoIter: Send,
     {
-        todo!()
+        let locked_db = self.lock_db().await?;
+        let db_transaction = locked_db.get_inner().transaction().await?;
+        let table = db_transaction.table::<NftTxHistoryTable>().await?;
+        for tx in txs {
+            let tx_item = NftTxHistoryTable::from_tx_history(&tx)?;
+            table.add_item(&tx_item).await?;
+        }
+        Ok(())
     }
 
-    async fn get_last_block_number(&self, _chain: &Chain) -> MmResult<Option<u32>, Self::Error> { todo!() }
+    async fn get_last_block_number(&self, chain: &Chain) -> MmResult<Option<u64>, Self::Error> {
+        let locked_db = self.lock_db().await?;
+        let db_transaction = locked_db.get_inner().transaction().await?;
+        let table = db_transaction.table::<NftTxHistoryTable>().await?;
+        let maybe_item = table
+            .cursor_builder()
+            .only("chain", chain.to_string())
+            .map_err(|e| WasmNftCacheError::GetLastNftBlockError(e.to_string()))?
+            .bound("block_number", BeBigUint::from(0u64), BeBigUint::from(u64::MAX))
+            .reverse()
+            .open_cursor(NftTxHistoryTable::CHAIN_BLOCK_NUMBER_INDEX)
+            .await
+            .map_err(|e| WasmNftCacheError::GetLastNftBlockError(e.to_string()))?
+            .next()
+            .await
+            .map_err(|e| WasmNftCacheError::GetLastNftBlockError(e.to_string()))?;
+        let maybe_item = maybe_item
+            .map(|(_, item)| {
+                item.block_number
+                    .to_u64()
+                    .ok_or_else(|| WasmNftCacheError::GetLastNftBlockError("height is too large".to_string()))
+            })
+            .transpose()?;
+        Ok(maybe_item)
+    }
```

Comment:
I see you have duplicate of this function with `IndexedDbNftStorage::get_last_block_number`  can you create a helper function and use once in both places

Reply:
Done